Pilot/pilotCommands.py

```diff
@@ -431,16 +451,58 @@ def _installDIRACpy3(self):
                 self.log.error("Could not pip install %s [ERROR %d]" % (self.releaseVersion, retCode))
                 self.exitWithError(retCode)
 
-    @logFinalizer
-    def execute(self):
-        """What is called all the time"""
+    def _localInstallDIRAC(self):
+        self.log.info("Installing DIRAC locally")
         if self.pp.pythonVersion == "27":
             self._setInstallOptions()
             self._locateInstallationScript()
-            self._installDIRAC()
+            self._installDIRACpy2()
         else:  # python3 is requested
             self._installDIRACpy3()
 
+    @logFinalizer
+    def execute(self):
+        """What is called all the time"""
+
+        try:
+            # In case we want to force local installation (in absence of CVMFS or for test reasons)
+            if "diracInstallOnly" in self.pp.genericOption:
+                self.log.info("NOT sourcing: starting traditional DIRAC installation")
+                self._localInstallDIRAC()
+            
+            else:
+                # Try sourcing from CVMFS
+                self._getPreinstalledEnvScript()
+                if self.pp.preinstalledEnv:
+                    self._sourceEnvironmentFile()
+                    self.log.info("source DIRAC env DONE, for release %s" % self.pp.releaseVersion)
+                    # environment variables to add?
+                    if self.pp.userEnvVariables:
+                        # User-requested environment variables (comma-separated, name and value separated by ":::")
+                        # The case of dirac-install is handled in vanilla DIRAC Pilot
+                        newEnvVars = dict(
+                            name.split(":::", 1) for name in self.pp.userEnvVariables.replace(" ", "").split(",")
+                        )
+                        self.log.info("Adding env variable(s) to the environment", newEnvVars)
+                        self.pp.installEnv.update(newEnvVars)
+                else: 
+                    self._localInstallDIRAC()
```

Comment:
```suggestion
            # In case we want to force local installation (in absence of CVMFS or for test reasons)
            if "diracInstallOnly" in self.pp.genericOption:
                self.log.info("NOT sourcing: starting traditional DIRAC installation")
                self._localInstallDIRAC()
                return
            
            # Try sourcing from CVMFS
            self._getPreinstalledEnvScript()
            if not self.pp.preinstalledEnv:
                self._localInstallDIRAC()
                return

            # if we are here, we have a preinstalled environment
            self._sourceEnvironmentFile()
            self.log.info("source DIRAC env DONE, for release %s" % self.pp.releaseVersion)
            # environment variables to add?
            if self.pp.userEnvVariables:
                # User-requested environment variables (comma-separated, name and value separated by ":::")
                # The case of dirac-install is handled in vanilla DIRAC Pilot
                newEnvVars = dict(
                name.split(":::", 1) for name in self.pp.userEnvVariables.replace(" ", "").split(",")
              )
                self.log.info("Adding env variable(s) to the environment", newEnvVars)
                self.pp.installEnv.update(newEnvVars)
```

Even with the `return` the `finally` block gets executed, so this gives a flatter code-path, which I find easier to reason about.

Reply:
I think the `else` is necessary here, otherwise we might end up installing twice.

Reply:
Which `else`? Did I miss a `return` or something else?

Reply:
PS: I could have also messed up the indentation, because I did that in the github comment window.

Reply:
Maybe I figured it out, check the version I just pushed.